tests/network/pod_networking.go

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/tests"
+	cd "kubevirt.io/kubevirt/tests/containerdisk"
+)
+
+var _ = SIGDescribe("Pod Networking", func() {
+	var virtClient kubecli.KubevirtClient
+
+	BeforeEach(func() {
+		var err error
+		virtClient, err = kubecli.GetKubevirtClient()
+		tests.PanicOnError(err)
+	})
+
+	vmiWithDefaultBinding := func() *v1.VirtualMachineInstance {
+		vmi := tests.NewRandomVMIWithEphemeralDisk(cd.ContainerDiskFor(cd.ContainerDiskAlpine))
+		vmi.Spec.Domain.Devices.Interfaces = nil
+		vmi.Spec.Networks = nil
+		return vmi
+	}
+
+	vmiWithBridgeBinding := func() *v1.VirtualMachineInstance {
+		vmi := tests.NewRandomVMIWithEphemeralDisk(cd.ContainerDiskFor(cd.ContainerDiskAlpine))
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{*v1.DefaultBridgeNetworkInterface()}
+		vmi.Spec.Networks = []v1.Network{*v1.DefaultPodNetwork()}
+		return vmi
+	}
+
+	vmiWithMasqueradeBinding := func() *v1.VirtualMachineInstance {
+		vmi := tests.NewRandomVMIWithEphemeralDisk(cd.ContainerDiskFor(cd.ContainerDiskAlpine))
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{*v1.DefaultMasqueradeNetworkInterface()}
+		vmi.Spec.Networks = []v1.Network{*v1.DefaultPodNetwork()}
+		return vmi
+	}
+
+	vmiWithBinding := func(bindingMechanism string) *v1.VirtualMachineInstance {
+		switch bindingMechanism {
+		case "bridge":
+			return vmiWithBridgeBinding()
+		case "masquerade":
+			return vmiWithMasqueradeBinding()
+		case "default":
+			return vmiWithDefaultBinding()
+		default:
+			panic(fmt.Sprintf("Selected binding %q is invalid", bindingMechanism))
+		}
+	}
+
+	testStatusReporting := func(description string, bindingMechanism string) {
```

Comment:
There is a format used in Ginkgo that is expected when reading the tests  and their report.
I find this wrapping surprising and unorthodox.

I have noticed http://onsi.github.io/ginkgo/#shared-example-patterns which introduces two patterns, both handling "it" and nothing else. I favor the 2nd pattern.

Ginkgo itself is pretty limited (too limited for my taste), but it seems we can use extensions like the [table](https://github.com/onsi/ginkgo/blob/master/extensions/table/table.go) one.

What is missing here is a parametrize ability for the fixtures. Do you think we can write an extension for that?

I would consider pattern 2 until then.

Reply:
I too find it little odd. I was playing with different context structures and templates but at the end found that this "odd" format may fit out use (heavy on parametrization and slow BeforeEach) the best.

See my follow-up draft that adds another suite and hopefully illustrates this better https://github.com/kubevirt/kubevirt/blob/dd788c799f97b3feea65fcdba7a97858cc8a650f/tests/network/pod_networking.go.

Regarding the patterns you shared, both of the options there have dedicated context per each test. That would mean that we have to duplicate it per every single parameter. I don't think that is a viable alternative for the "matrix" we need. We would be flooded by contexts.

What is missing in tables is that we cannot have AfterEach and BeforeEach in them. Since we don't want to use the pattern of "cleanup everything before each", we have a following issue:

```
Describe:
    TableContext:
        Test: create a VM and test that it works
        Entries:
          VMtypeA
          VMtypeB
```

Since the source of parametrization is table's entry we have to initialize the VM inside the test itself. That is bad on its own but it also means that we have no way to cleanup after the test is done and data collected.

I encourage you to look at the follow-up PR I linked. The closure pattern is not that different from those you referenced with the exception that it allows us to reuse parametrized Context. I believe it is a good fit for our tests compared to extremely verbose context repetition (not speaking about cases where the matrix will have 2 dimensions and more).

Reply:
Meh, lemme try opt 2 for the sake of discussion. Will add it at the top of the PR for now.

Reply:
Check it out.